File: packages/runar-rb/lib/runar/sdk/gorillapool_provider.rb
# frozen_string_literal: true

require 'net/http'
require 'json'
require 'uri'
require_relative 'provider'
require_relative 'types'

# GorillaPoolProvider -- HTTP-based BSV provider for 1sat Ordinals.
#
# Implements the Provider interface plus ordinal-specific methods for
# querying inscriptions, BSV-20/BSV-21 balances, and token UTXOs.
#
# Endpoints:
#   Mainnet: https://ordinals.gorillapool.io/api/
#   Testnet: https://testnet.ordinals.gorillapool.io/api/
#
#   provider = Runar::SDK::GorillaPoolProvider.new(network: 'mainnet')
#   utxos = provider.get_utxos('1A1zP1eP...')

module Runar
  module SDK
    class GorillaPoolProvider < Provider
      # @param network [String] 'mainnet' or 'testnet'
      def initialize(network: 'mainnet')
        @network = network
        @base_url = if network == 'mainnet'
                      'https://ordinals.gorillapool.io/api'
                    else
                      'https://testnet.ordinals.gorillapool.io/api'
                    end
      end

      # -------------------------------------------------------------------
      # Standard Provider methods
      # -------------------------------------------------------------------

      # Fetch a TransactionData by its txid.
      #
      # @param txid [String] transaction id (64 hex chars)
      # @return [TransactionData]
      def get_transaction(txid)
        data = api_get("/tx/#{txid}")

        inputs = Array(data['vin']).map do |vin|
          TxInput.new(
            txid: vin['txid'],
            output_index: vin['vout'],
            script: vin.dig('scriptSig', 'hex') || '',
            sequence: vin['sequence'] || 0xFFFFFFFF
          )
        end

        outputs = Array(data['vout']).map do |vout|
          satoshis = vout['value'].is_a?(Numeric) && vout['value'] < 1000 ? (vout['value'] * 1e8).round : vout['value']
          TxOutput.new(
            satoshis: satoshis,
            script: vout.dig('scriptPubKey', 'hex') || ''
          )
        end

        TransactionData.new(
          txid: data['txid'],
          version: data['version'] || 1,
          inputs: inputs,
          outputs: outputs,
          locktime: data['locktime'] || 0,
          raw: data['hex'] || ''
        )
      end

      # Broadcast a raw transaction hex to the network.
      #
      # @param raw_tx [String] hex-encoded raw transaction
      # @return [String] txid of the broadcasted transaction
      def broadcast(raw_tx)
        uri = URI("#{@base_url}/tx")
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true

        request = Net::HTTP::Post.new(uri)
        request['Content-Type'] = 'application/json'
        request.body = JSON.generate(rawTx: raw_tx)

        response = http.request(request)
        unless response.is_a?(Net::HTTPSuccess)
          raise "GorillaPool broadcast failed (#{response.code}): #{response.body}"
        end

        result = JSON.parse(response.body)
        result.is_a?(String) ? result : (result['txid'] || '')
      end

      # Return all UTXOs for a given address.
      #
      # @param address [String] BSV address
      # @return [Array<Utxo>]
      def get_utxos(address)
        entries = api_get("/address/#{address}/utxos")
        return [] unless entries.is_a?(Array)

        entries.map do |e|
          Utxo.new(
            txid: e['txid'],
            output_index: e['vout'],
            satoshis: e['satoshis'],
            script: e['script'] || ''
          )
        end
      rescue RuntimeError => e
        return [] if e.message.include?('404')

        raise
      end

      # Find a contract UTXO by its script hash.
      #
      # @param script_hash [String] hex-encoded script hash
      # @return [Utxo, nil]
      def get_contract_utxo(script_hash)
        entries = api_get("/script/#{script_hash}/utxos")
        return nil unless entries.is_a?(Array) && !entries.empty?

        first = entries[0]
        Utxo.new(
          txid: first['txid'],
          output_index: first['vout'],
          satoshis: first['satoshis'],
          script: first['script'] || ''
        )
      rescue RuntimeError => e
        return nil if e.message.include?('404')

        raise
      end

      # Return the network this provider is connected to.
      #
      # @return [String] 'mainnet' or 'testnet'
      def get_network
        @network
      end

      # Fetch the raw transaction hex by its txid.
      #
      # @param txid [String]
      # @return [String] hex-encoded raw transaction
      def get_raw_transaction(txid)
        uri = URI("#{@base_url}/tx/#{txid}/hex")
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true

        response = http.request(Net::HTTP::Get.new(uri))
        unless response.is_a?(Net::HTTPSuccess)
          raise "GorillaPool getRawTransaction failed (#{response.code}): #{response.body}"
        end

        response.body.strip
      end

      # Return the current fee rate in satoshis per kilobyte.
      #
      # BSV standard relay fee is 0.1 sat/byte (100 sat/KB).
      #
      # @return [Numeric]
      def get_fee_rate
        100
      end

      # -------------------------------------------------------------------
      # Ordinal-specific methods
      # -------------------------------------------------------------------

      # Get all inscriptions associated with an address.
      #
      # @param address [String]
      # @return [Array<Hash>] inscription info hashes
      def get_inscriptions_by_address(address)
        entries = api_get("/inscriptions/address/#{address}")
        return [] unless entries.is_a?(Array)

        entries
      rescue RuntimeError => e
        return [] if e.message.include?('404')

        raise
      end

      # Get inscription details (including content) by inscription ID.
      #
      # @param inscription_id [String] format: "<txid>_<vout>"
      # @return [Hash] inscription detail hash
      def get_inscription(inscription_id)
        api_get("/inscriptions/#{inscription_id}")
      end

      # Get BSV-20 (v1, tick-based) token balance for an address.
      #
      # @param address [String]
      # @param tick [String]
      # @return [String] balance
      def get_bsv20_balance(address, tick)
        result = api_get("/bsv20/balance/#{address}/#{URI.encode_www_form_component(tick)}")
        result.is_a?(String) ? result : (result['balance'] || '0').to_s
      rescue RuntimeError => e
        return '0' if e.message.include?('404')

        raise
      end

      # Get BSV-20 token UTXOs for an address and ticker.
      #
      # @param address [String]
      # @param tick [String]
      # @return [Array<Utxo>]
      def get_bsv20_utxos(address, tick)
        entries = api_get("/bsv20/utxos/#{address}/#{URI.encode_www_form_component(tick)}")
        return [] unless entries.is_a?(Array)

        entries.map do |e|
          Utxo.new(
            txid: e['txid'],
            output_index: e['vout'],
            satoshis: e['satoshis'],
            script: e['script'] || ''
          )
        end
      rescue RuntimeError => e
        return [] if e.message.include?('404')

        raise
      end

      # Get BSV-21 (v2, ID-based) token balance for an address.
      #
      # @param address [String]
      # @param id [String] token ID (format: "<txid>_<vout>")
      # @return [String] balance
      def get_bsv21_balance(address, id)
        result = api_get("/bsv20/balance/#{address}/#{URI.encode_www_form_component(id)}")
        result.is_a?(String) ? result : (result['balance'] || '0').to_s
      rescue RuntimeError => e
        return '0' if e.message.include?('404')

        raise
      end

      # Get BSV-21 token UTXOs for an address and token ID.
      #
      # @param address [String]
      # @param id [String] token ID (format: "<txid>_<vout>")
      # @return [Array<Utxo>]
      def get_bsv21_utxos(address, id)
        entries = api_get("/bsv20/utxos/#{address}/#{URI.encode_www_form_component(id)}")
        return [] unless entries.is_a?(Array)

        entries.map do |e|
          Utxo.new(
            txid: e['txid'],
            output_index: e['vout'],
            satoshis: e['satoshis'],
            script: e['script'] || ''
          )
        end
      rescue RuntimeError => e
        return [] if e.message.include?('404')

        raise
      end

      private

      # Perform an HTTP GET request against the GorillaPool API and parse the JSON response.
      #
      # @param path [String] API path (appended to base_url)
      # @return [Object] parsed JSON response
      def api_get(path)
        uri = URI("#{@base_url}#{path}")
        http = Net::HTTP.new(uri.host, uri.port)
        http.use_ssl = true

        response = http.request(Net::HTTP::Get.new(uri))
        unless response.is_a?(Net::HTTPSuccess)
          raise "GorillaPool request failed (#{response.code}): #{response.body}"
        end

        JSON.parse(response.body)
      end
    end
  end
end

File: packages/runar-java/src/main/java/runar/lang/sdk/GorillaPoolProvider.java
package runar.lang.sdk;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * REST {@link Provider} backed by the GorillaPool 1sat Ordinals API.
 * Parity with {@code packages/runar-go/sdk_gorillapool.go}.
 *
 * <p>Endpoints:
 * <ul>
 *     <li>Mainnet: {@code https://ordinals.gorillapool.io/api}</li>
 *     <li>Testnet: {@code https://testnet.ordinals.gorillapool.io/api}</li>
 * </ul>
 *
 * <p>UTXO endpoints return objects with {@code txid}, {@code vout},
 * {@code satoshis}, and (optional) {@code script} keys. The broadcast
 * endpoint accepts a {@code {"rawTx":"..."}} body and may return either a
 * JSON-encoded txid string or a {@code {"txid":"..."}} envelope; both
 * forms are tolerated.
 */
public final class GorillaPoolProvider implements Provider {

    private final String network;
    private final String baseUrl;
    private final HttpTransport transport;

    public GorillaPoolProvider() {
        this("mainnet", HttpTransport.jdkDefault());
    }

    public GorillaPoolProvider(String network) {
        this(network, HttpTransport.jdkDefault());
    }

    // Package-private full constructor so tests can inject a fake transport.
    GorillaPoolProvider(String network, HttpTransport transport) {
        String n = (network == null || network.isEmpty()) ? "mainnet" : network;
        this.network = n;
        this.baseUrl = "testnet".equals(n)
            ? "https://testnet.ordinals.gorillapool.io/api"
            : "https://ordinals.gorillapool.io/api";
        this.transport = transport == null ? HttpTransport.jdkDefault() : transport;
    }

    public String getNetwork() {
        return network;
    }

    public String getBaseUrl() {
        return baseUrl;
    }

    // ------------------------------------------------------------------
    // Provider
    // ------------------------------------------------------------------

    @Override
    public List<UTXO> listUtxos(String address) {
        HttpTransport.Response resp = transport.send(
            "GET", baseUrl + "/address/" + address + "/utxos", null, null);
        if (resp.statusCode() == 404) return List.of();
        requireOk(resp, "GorillaPool getUtxos");

        return parseUtxoList(resp.body(), "GorillaPool getUtxos");
    }

    @Override
    public String broadcastRaw(String txHex) {
        if (txHex == null || txHex.isEmpty()) {
            throw new IllegalArgumentException("GorillaPoolProvider.broadcastRaw: txHex is required");
        }
        Map<String, Object> payload = new LinkedHashMap<>();
        payload.put("rawTx", txHex);
        Map<String, String> headers = Map.of("Content-Type", "application/json");

        HttpTransport.Response resp = transport.send(
            "POST", baseUrl + "/tx", headers, JsonWriter.write(payload));
        requireOk(resp, "GorillaPool broadcast");

        String body = resp.body() == null ? "" : resp.body().trim();
        // Try JSON-encoded string first, then {"txid": "..."}, then raw text.
        try {
            Object parsed = Json.parse(body);
            if (parsed instanceof String s) return s;
            if (parsed instanceof Map<?, ?> m) {
                Object t = ((Map<?, ?>) m).get("txid");
                if (t instanceof String s && !s.isEmpty()) return s;
            }
        } catch (RuntimeException ignored) {
            // Fall through to plain-text body.
        }
        return body;
    }

    @Override
    public UTXO getUtxo(String txid, int vout) {
        HttpTransport.Response resp = transport.send(
            "GET", baseUrl + "/tx/" + txid, null, null);
        if (resp.statusCode() == 404) return null;
        requireOk(resp, "GorillaPool getTransaction");

        Map<String, Object> data = Json.asObject(parseJson(resp.body(), "GorillaPool getTransaction"));
        Object voutArr = data.get("vout");
        if (!(voutArr instanceof List<?> outs) || vout < 0 || vout >= outs.size()) return null;
        Map<String, Object> out = Json.asObject(outs.get(vout));
        long sats = btcOrSatToSatoshis(out.get("value"));
        String script = "";
        if (out.get("scriptPubKey") instanceof Map<?, ?> sp) {
            Object hex = ((Map<?, ?>) sp).get("hex");
            if (hex instanceof String s) script = s;
        }
        return new UTXO(txid, vout, sats, script);
    }

    /**
     * Fetches the raw transaction hex by txid via {@code /tx/{txid}/hex}.
     * Mirrors {@code GorillaPoolProvider.GetRawTransaction} in the Go SDK.
     */
    public String getRawTransaction(String txid) {
        HttpTransport.Response resp = transport.send(
            "GET", baseUrl + "/tx/" + txid + "/hex", null, null);
        requireOk(resp, "GorillaPool getRawTransaction");
        return resp.body() == null ? "" : resp.body().trim();
    }

    /**
     * Looks up unspent outputs for a script hash via
     * {@code /script/{scriptHash}/utxos}. Returns an empty list when the
     * upstream returns 404. Useful for stateful contract continuations.
     */
    public List<UTXO> getContractUtxos(String scriptHash) {
        HttpTransport.Response resp = transport.send(
            "GET", baseUrl + "/script/" + scriptHash + "/utxos", null, null);
        if (resp.statusCode() == 404) return List.of();
        requireOk(resp, "GorillaPool getContractUtxos");
        return parseUtxoList(resp.body(), "GorillaPool getContractUtxos");
    }

    // ------------------------------------------------------------------
    // Ordinal / token helpers (parity with Go SDK)
    // ------------------------------------------------------------------

    /** {@code /bsv20/balance/{address}/{tick}} — returns the balance string, or "0". */
    public String getBSV20Balance(String address, String tick) {
        return getTokenBalance(address, tick, "getBSV20Balance");
    }

    /** {@code /bsv20/utxos/{address}/{tick}}. */
    public List<UTXO> getBSV20Utxos(String address, String tick) {
        return getTokenUtxos(address, tick, "getBSV20Utxos");
    }

    /** {@code /bsv20/balance/{address}/{id}} — id is {@code <txid>_<vout>}. */
    public String getBSV21Balance(String address, String id) {
        return getTokenBalance(address, id, "getBSV21Balance");
    }

    /** {@code /bsv20/utxos/{address}/{id}} — id is {@code <txid>_<vout>}. */
    public List<UTXO> getBSV21Utxos(String address, String id) {
        return getTokenUtxos(address, id, "getBSV21Utxos");
    }

    private String getTokenBalance(String address, String key, String op) {
        HttpTransport.Response resp = transport.send(
            "GET", baseUrl + "/bsv20/balance/" + address + "/" + urlEncode(key), null, null);
        if (resp.statusCode() == 404) return "0";
        requireOk(resp, "GorillaPool " + op);

        String body = resp.body() == null ? "" : resp.body().trim();
        try {
            Object parsed = Json.parse(body);
            if (parsed instanceof String s) return s;
            if (parsed instanceof Map<?, ?> m) {
                Object bal = ((Map<?, ?>) m).get("balance");
                if (bal instanceof String s && !s.isEmpty()) return s;
            }
        } catch (RuntimeException ignored) {
            // Fall through.
        }
        return "0";
    }

    private List<UTXO> getTokenUtxos(String address, String key, String op) {
        HttpTransport.Response resp = transport.send(
            "GET", baseUrl + "/bsv20/utxos/" + address + "/" + urlEncode(key), null, null);
        if (resp.statusCode() == 404) return List.of();
        requireOk(resp, "GorillaPool " + op);
        return parseUtxoList(resp.body(), "GorillaPool " + op);
    }

    // ------------------------------------------------------------------
    // Helpers
    // ------------------------------------------------------------------

    private static List<UTXO> parseUtxoList(String body, String op) {
        Object parsed = parseJson(body, op);
        if (!(parsed instanceof List<?> entries)) {
            throw new ProviderException(op + ": expected JSON array, got " + typeName(parsed));
        }
        List<UTXO> utxos = new ArrayList<>(entries.size());
        for (Object e : entries) {
            Map<String, Object> u = Json.asObject(e);
            String txid = Json.asString(u.get("txid"));
            int vout = Json.asInt(u.get("vout"));
            long sats = Json.asLong(u.get("satoshis"));
            String script = "";
            if (u.get("script") instanceof String s) script = s;
            utxos.add(new UTXO(txid, vout, sats, script));
        }
        return utxos;
    }

    private static long btcOrSatToSatoshis(Object value) {
        if (value == null) return 0L;
        if (value instanceof Double d) {
            // GorillaPool may serialize as raw satoshis (integer-shaped doubles)
            // or as BTC. Mirror the Go SDK heuristic: < 1000 means BTC.
            if (d < 1000.0) return Math.round(d * 1e8);
            return d.longValue();
        }
        if (value instanceof Long l) return l;
        if (value instanceof Integer i) return i;
        if (value instanceof java.math.BigInteger bi) return bi.longValueExact();
        if (value instanceof String s) {
            double parsed = Double.parseDouble(s);
            if (parsed < 1000.0) return Math.round(parsed * 1e8);
            return (long) parsed;
        }
        throw new ProviderException("expected numeric vout.value, got " + typeName(value));
    }

    private static void requireOk(HttpTransport.Response resp, String op) {
        if (resp.statusCode() < 200 || resp.statusCode() >= 300) {
            throw new ProviderException(
                op + " failed (" + resp.statusCode() + "): "
                    + RPCProvider.truncate(resp.body(), 256),
                resp.statusCode());
        }
    }

    private static Object parseJson(String body, String op) {
        try {
            return Json.parse(body);
        } catch (RuntimeException re) {
            throw new ProviderException(op + " JSON decode failed: " + re.getMessage(), re);
        }
    }

    private static String urlEncode(String v) {
        return URLEncoder.encode(v, StandardCharsets.UTF_8);
    }

    private static String typeName(Object v) {
        return v == null ? "null" : v.getClass().getSimpleName();
    }
}
